### DISCUSSION TITLE
Enable replacing powershell.exe and 5.1 SMA.dll in Windows

### DISCUSSION THREAD

**Discussion by SteveL-MSFT** [PowerShell/PowerShell#20789] [Enhancements]
Many applications (and inbox components of Windows) use powershell.exe or host SMA.dll so there's no way currently to have them try to use PS7 instead.

- For powershell.exe, it would be possible to have a `powershell.exe` that spawns `pwsh` as passes the args, but any application that uses the full path under System32, this won't work.
- For applications hosting SMA.dll, we can't simply replace it with the PS7 one since a .NET Framework application wouldn't be able to host the .NET version of SMA.dll.  Instead, it may be feasible to have a SMA.dll shim using IPC, but the objects would be deserialized

**Comment by rhubarb-geek-nz** [comment]
Can you describe a scenario?

If an application has been built to use classic PowerShell as an integrated part of Windows, why do you want it to use "PowerShell Core", I am presuming all its scripts have been tested etc against classic PowerShell so changing it is likely to break things.

In order to use a component it does not mean it has to run in the same process, so a .NET Framework application can happily use PowerShell Core by running it in it's own process, just like 64 bit and 32 bit processes can co-exist on the same machine.

**Comment by kasini3000** [comment]
new a pwsh_SMA.dll ? ,then let psversion gt 7  use it .

**Comment by fMichaleczek** [comment]
I advocate for a comprehensive approach from Microsoft. Rather than attempting to replace PowerShell.exe with pwsh.exe, a more practical strategy involves developing a new Windows Management Framework. This framework should address key concerns for script developers, incorporating substantial enhancements and fixes.

Simultaneously, unifying the package manager by adding support for (nuget) libraries based on framework versions would streamline the development process. Integrating this improved package manager into the system's inbox would enhance overall efficiency, providing a seamless experience for both PowerShell and Windows PowerShell users. 

Rather than solely focusing on technical transitions like the migration to the new dotnet framework, investing in the revitalization of the Windows Management Framework, and refining the existing Windows PowerShell environment would be a more pragmatic and beneficial approach. This comprehensive strategy aims to meet the needs of script developers and IT professionals, ensuring a unified, efficient, and user-friendly experience.
